File: packages/api-models-py/skyportal_py_models/invitations.py
"""Request and response models for ``/api/invitations``."""

from __future__ import annotations

import datetime

from pydantic import BaseModel, ConfigDict, Field

from skyportal_py_models.groups import Group
from skyportal_py_models.roles import Role
from skyportal_py_models.streams import Stream
from skyportal_py_models.users import User


class Invitation(BaseModel):
    """An invitation for a new user to join the instance (upstream ``Invitation``)."""

    # The handler eager-loads ``groups``, ``streams`` and ``invited_by``;
    # ``role`` is only present when that relationship happens to be loaded.

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    token: str | None = None
    user_email: str | None = None
    role_id: str | None = None
    role: Role | None = None
    admin_for_groups: list[bool] | None = None
    can_save_to_groups: list[bool] | None = None
    can_share_photometry_for_groups: list[bool] | None = None
    used: bool | None = None
    user_expiration_date: datetime.datetime | None = None
    groups: list[Group] | None = None
    streams: list[Stream] | None = None
    invited_by: User | None = None


class InvitationsPage(BaseModel):
    """One page of results from an invitations query."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    invitations: list[Invitation] = Field(default_factory=list)
    total_matches: int = Field(alias="totalMatches", default=0)


class InvitationPost(BaseModel):
    """Payload for inviting a new user."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    user_email: str = Field(alias="userEmail")
    group_ids: list[int] = Field(alias="groupIDs")
    role: str | None = None
    stream_ids: list[int] | None = Field(alias="streamIDs", default=None)
    group_admin: list[bool] | None = Field(alias="groupAdmin", default=None)
    can_save: list[bool] | None = Field(alias="canSave", default=None)
    can_share_photometry: list[bool] | None = Field(
        alias="canSharePhotometry", default=None
    )
    user_expiration_date: str | None = Field(alias="userExpirationDate", default=None)


class InvitationPostResponse(BaseModel):
    """Result of creating an invitation."""

    model_config = ConfigDict(extra="forbid")

    id: int

File: packages/api-models-ts/src/Invitations.ts
/**
 * Request and response models for `/api/invitations`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Groups from "./Groups.ts";
import * as Roles from "./Roles.ts";
import * as Schemas from "./Schemas.ts";
import * as Streams from "./Streams.ts";
import * as Users from "./Users.ts";

/**
 * An invitation for a new user to join the instance (upstream `Invitation`).
 *
 * The handler eager-loads `groups`, `streams` and `invited_by`; `role` is only
 * present when that relationship happens to be loaded.
 *
 * @since 1.0.0
 * @category Models
 */
export const Invitation = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        token: Schemas.NullishString,
        user_email: Schemas.NullishString,
        role_id: Schemas.NullishString,
        role: Schemas.nullish(Roles.Role),
        admin_for_groups: Schemas.nullish(v.array(v.boolean())),
        can_save_to_groups: Schemas.nullish(v.array(v.boolean())),
        can_share_photometry_for_groups: Schemas.nullish(v.array(v.boolean())),
        used: Schemas.NullishBoolean,
        user_expiration_date: Schemas.NullishTimestamp,
        groups: Schemas.nullish(v.array(Groups.Group)),
        streams: Schemas.nullish(v.array(Streams.Stream)),
        invited_by: Schemas.nullish(Users.User),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type Invitation = v.InferOutput<typeof Invitation>;

/**
 * One page of results from an invitations query.
 *
 * @since 1.0.0
 * @category Models
 */
export const InvitationsPage = Schemas.model(
    v.strictObject({
        invitations: Schemas.list(Invitation),
        totalMatches: v.optional(Schemas.Integer, 0),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type InvitationsPage = v.InferOutput<typeof InvitationsPage>;

/**
 * Payload for inviting a new user.
 *
 * `role` must be either `"Full user"` or `"View only"` and defaults to `"Full
 * user"`. If `streamIDs` is omitted the user is granted access to every stream
 * associated with the invited groups; if given, it must cover those streams.
 * `groupAdmin`, `canSave` and `canSharePhotometry` are per-group flags and
 * must be the same length as `groupIDs`; they default to all false, all true,
 * and all false respectively.
 *
 * @since 1.0.0
 * @category Models
 */
export interface InvitationPost {
    readonly userEmail: string;
    readonly groupIDs: ReadonlyArray<number>;
    readonly role?: string | undefined;
    readonly streamIDs?: ReadonlyArray<number> | undefined;
    readonly groupAdmin?: ReadonlyArray<boolean> | undefined;
    readonly canSave?: ReadonlyArray<boolean> | undefined;
    readonly canSharePhotometry?: ReadonlyArray<boolean> | undefined;
    readonly userExpirationDate?: string | undefined;
}

/**
 * Result of creating an invitation.
 *
 * @since 1.0.0
 * @category Models
 */
export const InvitationPostResponse = Schemas.model(
    v.strictObject({ id: Schemas.Integer })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type InvitationPostResponse = v.InferOutput<typeof InvitationPostResponse>;
